give_kudos.py

```diff
@@ -38,52 +37,67 @@ def email_login(self):
         self.page.fill("#password", self.PASSWORD)
         self.page.click("button[type='submit']")
         print("---Logged in!!---")
-        self.page.goto(os.path.join(BASE_URL, "dashboard"), wait_until="domcontentloaded")
-        
+        # limit activities count by GET parameter
+        self.page.goto(os.path.join(BASE_URL, "dashboard?num_entries=50"), wait_until="networkidle")
+        self.own_profile_id = self.page.locator("#athlete-profile .card-body > a").get_attribute('href').split("/athletes/")[1]
+        print(f"own_profile_id: {self.own_profile_id}")
+
 
-    def locate_kudos_buttons_and_maybe_give_kudos(self, button_locator) -> int:
+    def locate_kudos_buttons_and_maybe_give_kudos(self, web_feed_entry_locator) -> int:
         """
         input: playwright.locator class
         Returns count of kudos given.
         """
-        b_count= button_locator.count()
+        w_count = web_feed_entry_locator.count()
         given_count = 0
-        for i in range(b_count):
-            button = button_locator.nth(i)
-            if button.get_by_test_id("unfilled_kudos").count():
-                try:
-                    button.click(timeout=2000)
-                    given_count += 1
-                    time.sleep(1)
-                except TimeoutError:
-                    print("Click timed out.")
+        print(f"web feeds found: {w_count}")
+        for i in range(w_count):
+            # run condition check
+            curr_duration = time.time() - self.start_time
+            if curr_duration > self.max_run_duration:
+                print("Max run duration reached.")
+                break
+
+            web_feed = web_feed_entry_locator.nth(i)
+            p_count = web_feed.get_by_test_id("entry-header").count()
+
+            # check if activity has multiple participants
+            if p_count > 1:
+                print(f"Found multiple list entries: {p_count}")
+                for j in range(p_count):
+                    participant = web_feed.get_by_test_id("entry-header").nth(j)
+                    # ignore own activities
+                    if participant.get_by_test_id("owners-name").get_attribute('href').split("/athletes/")[1] != self.own_profile_id:
+                        kudos_container = web_feed.get_by_test_id("kudos_comments_container").nth(j)
+                        button = kudos_container.get_by_test_id("unfilled_kudos")
+                        if button.count() == 1:
+                            print("Feed has no kudo")
+                            button.click(timeout=0, no_wait_after=True)
+                            print("Kudos button clicked")
+                            given_count += 1
+                            time.sleep(1)
+            else:
+                # ignore own activities
+                if web_feed.get_by_test_id("owners-name").get_attribute('href').split("/athletes/")[1] != self.own_profile_id:

```

Comment:
When I ran this on my local machine, the script seemed to have timed out on this line. We could wrap this in its own function with try/except, such that exceptions are caught. Default of this function can return an empty string.
```
Traceback (most recent call last):
  File "give_kudos.py", line 109, in <module>
    main()
  File "give_kudos.py", line 105, in main
    kg.give_kudos()
  File "give_kudos.py", line 98, in give_kudos
    self.locate_kudos_buttons_and_maybe_give_kudos(web_feed_entry_locator=web_feed_entry_locator)
  File "give_kudos.py", line 81, in locate_kudos_buttons_and_maybe_give_kudos
    if web_feed.get_by_test_id("owners-name").get_attribute('href').split("/athletes/")[1] != self.own_profile_id:
...
playwright._impl._api_types.TimeoutError: Timeout 30000ms exceeded.
=========================== logs ===========================
waiting for selector "[data-testid=web-feed-entry] >> nth=9 >> internal:attr=[data-testid="owners-name"]"
============================================================
```

Reply:
I run my version of the script since a week and only faced timeouts after login because my repeat intervalls were to short (got reCaptcha after login). Never saw timeouts at this line.

Reply:
But strange...

I have to use playwright 1.32.1 since 1.27.1 wasn't available for me on my system via pip, maybe this could be an issue locally on your side?

Reply:
It could be a version difference.